core/tx_multicurrency_priceheap.go

```diff
@@ -0,0 +1,134 @@
+package core
+
+import (
+	"container/heap"
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/core/types"
+)
+
+// multiCurrencyPriceHeap is a heap.Interface implementation over transactions
+// with different fee currencies for retrieving price-sorted transactions to discard
+// when the pool fills up. If baseFee is set then the heap is sorted based on the
+// effective tip based on the given base fee. If baseFee is nil then the sorting
+// is based on gasFeeCap.
+type multiCurrencyPriceHeap struct {
+	currencyCmp         CurrencyCmpFn
+	gpm                 GasPriceMinimums              // heap should always be re-sorted after gas price minimums (baseFees) is changed
+	nonNilCurrencyHeaps map[common.Address]*priceHeap // Heap of prices of all the stored non-nil currency transactions
```

Comment:
maybe we should rename to `nativeCurrency` and `nonNative`. `nil` is not really a thing

Reply:
I like it